src/stk/math/types.h

```diff
@@ -131,50 +132,97 @@ struct STK_ALIGN(16) double4
 #endif // STK_USE_CUDA
 
 
-template<typename T, unsigned int rows_, unsigned int cols_>
-struct matrix
+struct Matrix3x3f
 {
-    T _data[rows_][cols_];
+    static constexpr unsigned int rows = 3;
+    static constexpr unsigned int cols = 3;
 
-    static constexpr unsigned int rows = rows_;
-    static constexpr unsigned int cols = cols_;
+    float3 _rows[rows];
 
-    T* operator[](const unsigned int i) {
-        return _data[i];
+    float3& operator[](const unsigned int i) {
+        return _rows[i];
     }
 
-    const T* operator[](const unsigned int i) const {
-        return _data[i];
+    const float3& operator[](const unsigned int i) const {
+        return _rows[i];
     }
 
-    T& operator()(const unsigned int r, const unsigned int c) {
-        return _data[r][c];
+    const float& operator()(const unsigned int r, const unsigned int c) const {
+        float float3::*column;
+        switch (c)
+        {
+        case 0:
+            column = &float3::x;
+            break;
+        case 1:
+            column = &float3::y;
+            break;
+        case 2:
+            column = &float3::z;
+            break;
+        default:
+            FATAL() << "Index c = '" << c << "' is out of bound.";
+            break;
+        }
+        return _rows[r].*column;
+    }
+
+    float& operator()(const unsigned int r, const unsigned int c) {
+        return const_cast<float&>(static_cast<const Matrix3x3f*>(this)->operator()(r, c));
     }
 
-    const T& operator()(const unsigned int r, const unsigned int c) const {
-        return _data[r][c];
+    float* data(void) {
+        return reinterpret_cast<float*>(&_rows[0]);
     }
 
-    T* data(void) {
-        &_data[0][0];
+    const float* data(void) const {
+        return reinterpret_cast<const float*>(&_rows[0]);
     }
 
-    const T* data(void) const {
-        &_data[0][0];
+    void set(const float *data) {
+        std::copy(data, data + rows*cols, (float*) _rows);
     }
 
-    void diagonal(const std::initializer_list<T> d) {
-        ASSERT(d.size() == std::min(rows_, cols_));
-        std::fill(&_data[0][0], &_data[0][0] + rows_ * cols_, T(0));
+    void set(const std::initializer_list<float> data) {
+        std::copy(data.begin(), data.end(), (float*) _rows);
+    }
+
+    void diagonal(const std::initializer_list<float> d) {
+        ASSERT(d.size() == std::min(rows, cols));
+        std::fill(data(), data() + rows * cols, float(0));
         int i = 0;
         for (const auto& x : d) {
-            _data[i][i] = x;
+            (*this)(i, i) = x;
             i++;
         }
     }
-};
 
-using Matrix3x3f = matrix<float, 3, 3>;
+    float determinant(void) const {
+        return (*this)(0, 0) * ((*this)(1, 1) * (*this)(2, 2) - (*this)(1, 2) * (*this)(2, 1)) -
```

Comment:
Same here and in the inverse, should not be critical pieces, so I tried to make them a bit easier to read. Can't be sure before checking the assembly, but I think most of this stuff will be inlined however, so performance-wise it should not be too bad.

Reply:
It's only 3x3 so unless we do it _a lot_ this is probably fine. 😃

Reply:
I should check the assembly before talking, but I am quite confident the calls to `operator()` will be inlined, since they boil down to a memory access with `constexpr` offset. I was thinking to rewrite it using `_rows` directly, but it would make things unreadable.

In any case, the determinant should be used just for matrix inversion, which in turn is done only once when the direction matrix is set.